src/Forms/DatetimeField.php

```diff
@@ -58,59 +85,204 @@ class DatetimeField extends FormField
+        // See https://html.spec.whatwg.org/multipage/infrastructure.html#times
+        if ($timestamp === false && $this->getHTML5()) {
+            $fromFormatter->setPattern(str_replace(':ss', '', $fromFormatter->getPattern()));
+            $timestamp = $fromFormatter->parse($datetime);
+        }
+
+        if ($timestamp === false) {
+            return null;
+        }
+        return $toFormatter->format($timestamp) ?: null;
+    }
+
+    /**
+     * Get date formatter with the standard locale / date format
+     *
+     * @throws \LogicException
+     * @return IntlDateFormatter
+     */
+    protected function getFormatter()
+    {
+        if ($this->getHTML5() && $this->datetimeFormat && $this->datetimeFormat !== DBDatetime::ISO_DATETIME_NORMALISED) {
+            throw new \LogicException(
+                'Please opt-out of HTML5 processing of ISO 8601 dates via setHTML5(false) if using setDatetimeFormat()'
+            );
+        }
+
+        if ($this->getHTML5() && $this->dateLength) {
+            throw new \LogicException(
+                'Please opt-out of HTML5 processing of ISO 8601 dates via setHTML5(false) if using setDateLength()'
+            );
+        }
+
+        if ($this->getHTML5() && $this->locale) {
+            throw new \LogicException(
+                'Please opt-out of HTML5 processing of ISO 8601 dates via setHTML5(false) if using setLocale()'
+            );
+        }
+
+        $formatter = IntlDateFormatter::create(
+            $this->getLocale(),
+            $this->getDateLength(),
+            $this->getTimeLength(),
+            $this->getTimezone()
+        );
 
-        // Combine date components back into iso8601 string for the root value
-        $this->value = $this->dataValue();
+        if ($this->getHTML5()) {
+            // Browsers expect ISO 8601 dates, localisation is handled on the client.
+            // Add 'T' date and time separator to create W3C compliant format
+            $formatter->setPattern(DBDatetime::ISO_DATETIME_NORMALISED);
+        } elseif ($this->datetimeFormat) {
+            // Don't invoke getDatetimeFormat() directly to avoid infinite loop
+            $ok = $formatter->setPattern($this->datetimeFormat);
+            if (!$ok) {
+                throw new InvalidArgumentException("Invalid date format {$this->datetimeFormat}");
+            }
+        }
+        return $formatter;
+    }
+
+    /**
+     * Get date format in CLDR standard format
+     *
+     * This can be set explicitly. If not, this will be generated from the current locale
+     * with the current date length.
+     *
+     * @see http://userguide.icu-project.org/formatparse/datetime#TOC-Date-Field-Symbol-Table
+     */
+    public function getDatetimeFormat()
+    {
+        if ($this->getHTML5()) {
+            // Browsers expect ISO 8601 dates, localisation is handled on the client
+            $this->setDatetimeFormat(DBDatetime::ISO_DATETIME_NORMALISED);
```

Comment:
Getters shouldn't modify state in unnecessary ways. Just return the constant directly here.